src/arithmetic/list_funcs/list_funcs.c

```diff
@@ -22,14 +25,33 @@ SIValue AR_TOLIST(SIValue *argv, int argc) {
 	return array;
 }
 
-/* Returns a value in a specific index in an array.
+/* If given an array, returns a value in a specific index in an array.
    Valid index range is [-arrayLen, arrayLen).
    Invalid index will return null.
-   "RETURN [1, 2, 3][0]" will yield 1. */
+   "RETURN [1, 2, 3][0]" will yield 1.
+
+   If given a map or graph entity, returns the property value associated
+   with the given key string. */
 SIValue AR_SUBSCRIPT(SIValue *argv, int argc) {
 	ASSERT(argc == 2);
 	if(SI_TYPE(argv[0]) == T_NULL || SI_TYPE(argv[1]) == T_NULL) return SI_NullVal();
+	if(SI_TYPE(argv[0]) & (T_MAP | SI_GRAPHENTITY)) {
+		ASSERT(SI_TYPE(argv[1]) == T_STRING);
+		/* If the first argument is a map or graph entity, this is a property lookup of a form like:
+		 * WITH {val: 5} AS a return a['val']
+		 * MATCH (a) RETURN a['val']
+		 * Pass the arguments to the AR_PROPERTY function. */
+		SIValue property_args[3] = {argv[0], argv[1], SI_LongVal(ATTRIBUTE_NOTFOUND)};
+		return AR_PROPERTY(property_args, 3);
+	}
+
 	ASSERT(SI_TYPE(argv[0]) == T_ARRAY && SI_TYPE(argv[1]) == T_INT64);
```

Comment:
Doesn't go hand in hand with the following line, consider removing this `ASSERT`